packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
  * @param compatBehavior - Typed redundantly with CompatModeBehavior to ensure handling is added when updating that type
  */
 function compatBehaviorAllowsMessageType(
-	_unknownContainerRuntimeMessageType: never,
+	_unknownContainerRuntimeMessageType: UnknownContainerMessageType,
 	compatBehavior: "Ignore" | "FailToProcess" | undefined,
 ): boolean {
 	// undefined defaults to same behavior as "FailToProcess"
 	return compatBehavior === "Ignore";
 }
 
 /**
- * The unpacked runtime message / details to be handled or dispatched by the ContainerRuntime
+ * The unpacked runtime message / details to be handled or dispatched by the ContainerRuntime.
+ * Message type are differentiated via a `type` string and contain different contents depending on their type.
  *
  * IMPORTANT: when creating one to be serialized, set the properties in the order they appear here.
  * This way stringified values can be compared.
  */
-export interface ContainerRuntimeMessage {
+export interface ContainerRuntimeMessage<TType, TContents> {
 	/** Type of the op, within the ContainerRuntime's domain */
-	type: ContainerMessageType;
+	type: TType;
 	/** Domain-specific contents, interpreted according to the type */
-	contents: any;
+	contents: TContents;
 	/** Info describing how to handle this op in case the type is unrecognized (default: fail to process) */
 	compatDetails?: IContainerRuntimeMessageCompatDetails;
 }
 
+export type ContainerRuntimeDataStoreOpMessage = ContainerRuntimeMessage<
```

Comment:
Let's put all these types (and the existing related types/code) in a separate file, either in this PR or I can in a follow-up.

Reply:
Good idea, I can do it in this PR.

Reply:
They're now in `messageTypes.ts`